state-chain/pallets/cf-ingress-egress/src/boost_pool/tests.rs

```diff
@@ -0,0 +1,427 @@
+	pool.add_funds(BOOSTER_2, 1000);
+
+	assert_eq!(pool.provide_funds_for_boosting(BOOST_1, 500), Ok((500, 10)));
+	check_pool(&pool, [(BOOSTER_1, 755), (BOOSTER_2, 755)]);
+	check_pending_boosts(&pool, [(BOOST_1, vec![(BOOSTER_1, 250), (BOOSTER_2, 250)])]);
+
+	assert_eq!(pool.stop_boosting(BOOSTER_1), Ok(755));
+	check_pool(&pool, [(BOOSTER_2, 755)]);
+
+	// New booster joins while we have a pending withdrawal:
+	pool.add_funds(BOOSTER_3, 2000);
+	check_pool(&pool, [(BOOSTER_2, 755), (BOOSTER_3, 2000)]);
+
+	// The amount used for boosting from a given booster is proportional
+	// to their share in the available pool:
+	assert_eq!(pool.provide_funds_for_boosting(BOOST_2, 1000), Ok((1000, 20)));
+	check_pool(&pool, [(BOOSTER_2, 486), (BOOSTER_3, 1288)]);
+	check_pending_boosts(
+		&pool,
+		[
+			(BOOST_1, vec![(BOOSTER_1, 250), (BOOSTER_2, 250)]),
+			(BOOST_2, vec![(BOOSTER_2, 274), (BOOSTER_3, 725)]),
+		],
+	);
+
+	// Deposit of 500 is finalised, 250 goes to Booster 1's free balance, and the
+	// remaining 250 goes to Booster 2; Booster 3 joined after this boost, so they
+	// get nothing; there is only one pending boost now (Boost 2):
+	assert_eq!(pool.on_finalised_deposit(BOOST_1), vec![(BOOSTER_1, 250)]);
+	check_pool(&pool, [(BOOSTER_2, 736), (BOOSTER_3, 1288)]);
+	check_pending_boosts(&pool, [(BOOST_2, vec![(BOOSTER_2, 274), (BOOSTER_3, 725)])]);
+
+	{
+		// Scenario A: the second deposit is lost; available amounts remain the same,
+		// but there is no more pending boosts:
+		let mut pool = pool.clone();
+		pool.on_lost_deposit(BOOST_2);
+		check_pool(&pool, [(BOOSTER_2, 736), (BOOSTER_3, 1288)]);
+		check_pending_boosts(&pool, []);
+	}
+
+	{
+		// Scenario B: the second deposit is received and distributed back between
+		// the contributed boosters:
+		let mut pool = pool.clone();
+		assert_eq!(pool.on_finalised_deposit(BOOST_2), vec![]);
+		check_pool(&pool, [(BOOSTER_2, 1010), (BOOSTER_3, 2014)]);
+		check_pending_boosts(&pool, []);
+	}
+}
+
+/// Check that boosters with small contributions can boost can earn rewards that
+/// can be accumulated to non-zero chain amounts
+#[test]
+fn small_rewards_accumulate() {
+	// Booster 2 only owns a small fraction of the pool:
+	let mut pool = TestPool::new(100);
+	pool.add_funds(BOOSTER_1, 1000);
+	pool.add_funds(BOOSTER_2, 50);
+
+	const SMALL_DEPOSIT: AssetAmount = 500;
+
+	assert_eq!(pool.provide_funds_for_boosting(BOOST_1, SMALL_DEPOSIT), Ok((SMALL_DEPOSIT, 5)));
+	assert_eq!(pool.on_finalised_deposit(BOOST_1), vec![]);
+
+	// BOOSTER 2 earns ~0.25 (it is rounded down when converted to AssetAmount,
+	// but the fractional part isn't lost)
+	check_pool(&pool, [(BOOSTER_1, 1004), (BOOSTER_2, 50)]);
+
+	// 4 more boost like that and BOOSTER 2 should have withdrawable fees:
+	for boost_id in 1..=4 {
+		assert_eq!(
+			pool.provide_funds_for_boosting(boost_id, SMALL_DEPOSIT),
+			Ok((SMALL_DEPOSIT, 5))
+		);
+		assert_eq!(pool.on_finalised_deposit(boost_id), vec![]);
+	}
```

Comment:
Should we also check that the balance hasn't increased between these deposits?

Reply:
What's important here is that fractional rewards don't get lost, and not that it takes exactly 4 boosts for them to get 1 in the end. Adding more checks will only be a distractions IMO, especially since I would need to either figure out how balance of Booster 1 changes each iteration, or ingore it (but that's also not ideal).